File: data_worker/src/services/authcontrol/auth_authority_bootstrap_rpc_client.py
from __future__ import annotations

import asyncio
import base64
import time
from dataclasses import dataclass

import grpc
from google.protobuf import json_format
from google.protobuf import struct_pb2


BOOTSTRAP_INIT_METHOD = "/bms.auth.v1.AuthAuthorityBootstrapService/InitBootstrapChallenge"
BOOTSTRAP_AUTH_METHOD = "/bms.auth.v1.AuthAuthorityBootstrapService/AuthenticateBootstrap"


@dataclass(slots=True, kw_only=True)
class BootstrapHandshakeResult:
    stage: str
    active_comm_key_id: str = ""


class AuthAuthorityBootstrapRPCClient:
    def __init__(
        self,
        endpoint: str,
        *,
        dial_timeout_sec: float = 3.0,
        call_timeout_sec: float = 5.0,
    ) -> None:
        self._endpoint = (endpoint or "").strip()
        self._dial_timeout_sec = dial_timeout_sec
        self._call_timeout_sec = call_timeout_sec

    async def execute_bootstrap_handshake(
        self,
        *,
        entity_type: str,
        entity_id: str,
        audience: str,
        key_id: str,
    ) -> BootstrapHandshakeResult:
        if not self._endpoint:
            raise ValueError("auth authority endpoint is required")

        payload = {
            "entity_type": (entity_type or "").strip(),
            "entity_id": (entity_id or "").strip(),
            "audience": (audience or "").strip(),
            "key_id": (key_id or "").strip(),
            "ttl_sec": 60,
        }
        if not payload["entity_type"] or not payload["entity_id"]:
            raise ValueError("bootstrap entity_type and entity_id are required")

        async with grpc.aio.insecure_channel(self._endpoint) as channel:
            await asyncio.wait_for(
                channel.channel_ready(),
                timeout=self._dial_timeout_sec,
            )

            init_call = channel.unary_unary(
                BOOTSTRAP_INIT_METHOD,
                request_serializer=_serialize_struct,
                response_deserializer=_deserialize_struct,
            )
            challenge_resp = await init_call(
                _dict_to_struct(payload),
                timeout=self._call_timeout_sec,
            )
            challenge_envelope = _struct_to_dict(challenge_resp)
            challenge = challenge_envelope.get("challenge")
            if not isinstance(challenge, dict):
                challenge = challenge_envelope
            challenge_id = str(challenge.get("challenge_id", "")).strip()
            challenge_key_id = str(challenge.get("key_id", "")).strip() or payload["key_id"]
            if not challenge_id:
                raise RuntimeError("bootstrap challenge response missing challenge_id")

            auth_call = channel.unary_unary(
                BOOTSTRAP_AUTH_METHOD,
                request_serializer=_serialize_struct,
                response_deserializer=_deserialize_struct,
            )
            auth_req = {
                "challenge": challenge,
                "signed": {
                    "challenge_id": challenge_id,
                    "key_id": challenge_key_id,
                    "signature_algorithm": "ed25519",
                    "signature": base64.b64encode(
                        f"bootstrap:{challenge_id}".encode("utf-8")
                    ).decode("utf-8"),
                    "signed_at_ms": int(time.time() * 1000),
                },
                "scopes": ["service:bootstrap"],
                "role": "service",
                "require_downstream_token": False,
            }
            auth_resp = await auth_call(
                _dict_to_struct(auth_req),
                timeout=self._call_timeout_sec,
            )

        auth_payload = _struct_to_dict(auth_resp)
        stage = _normalize_bootstrap_stage(str(auth_payload.get("stage", "")).strip())
        if not stage:
            raise RuntimeError("bootstrap authenticate response missing stage")
        return BootstrapHandshakeResult(
            stage=stage,
            active_comm_key_id=str(auth_payload.get("active_comm_key_id", "")).strip(),
        )


def _dict_to_struct(payload: dict) -> struct_pb2.Struct:
    msg = struct_pb2.Struct()
    msg.update(payload)
    return msg


def _struct_to_dict(payload: struct_pb2.Struct) -> dict:
    return json_format.MessageToDict(payload, preserving_proto_field_name=True)


def _serialize_struct(payload: struct_pb2.Struct) -> bytes:
    return payload.SerializeToString()


def _deserialize_struct(raw: bytes) -> struct_pb2.Struct:
    msg = struct_pb2.Struct()
    msg.ParseFromString(raw)
    return msg


def _normalize_bootstrap_stage(raw: str) -> str:
    stage = (raw or "").strip().lower()
    if stage in {"ready", "bootstrap_stage_ready", "4"}:
        return "ready"
    if stage in {"uninitialized", "bootstrap_stage_uninitialized", "1"}:
        return "uninitialized"
    if stage in {"challenging", "bootstrap_stage_challenging", "2"}:
        return "challenging"
    if stage in {"authenticating", "bootstrap_stage_authenticating", "3"}:
        return "authenticating"
    return stage

File: gateway/src/services/authcontrol/auth_authority_bootstrap_rpc_client.go
package authcontrol

import (
	"context"
	"encoding/base64"
	"fmt"
	"strings"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/types/known/structpb"
)

const (
	bootstrapInitMethodPath = "/bms.auth.v1.AuthAuthorityBootstrapService/InitBootstrapChallenge"
	bootstrapAuthMethodPath = "/bms.auth.v1.AuthAuthorityBootstrapService/AuthenticateBootstrap"
)

// BootstrapHandshakeRequest 表示网关在冷启动阶段向认证中心发起的最小握手请求。
type BootstrapHandshakeRequest struct {
	EntityType string
	EntityID   string
	Audience   string
	KeyID      string
}

// BootstrapHandshakeResult 表示最小握手结果快照。
type BootstrapHandshakeResult struct {
	Stage           string
	ActiveCommKeyID string
}

// BootstrapRPCClient 负责执行认证中心 bootstrap 最小 RPC 调用。
type BootstrapRPCClient struct {
	endpoint    string
	dialTimeout time.Duration
	callTimeout time.Duration
}

func NewBootstrapRPCClient(endpoint string) *BootstrapRPCClient {
	return &BootstrapRPCClient{
		endpoint:    strings.TrimSpace(endpoint),
		dialTimeout: 3 * time.Second,
		callTimeout: 5 * time.Second,
	}
}

func (c *BootstrapRPCClient) ExecuteBootstrapHandshake(
	ctx context.Context,
	req *BootstrapHandshakeRequest,
) (*BootstrapHandshakeResult, error) {
	if req == nil {
		return nil, fmt.Errorf("bootstrap handshake request is nil")
	}
	if c.endpoint == "" {
		return nil, fmt.Errorf("auth authority endpoint is required")
	}
	if strings.TrimSpace(req.EntityType) == "" || strings.TrimSpace(req.EntityID) == "" {
		return nil, fmt.Errorf("bootstrap entity_type and entity_id are required")
	}

	dialCtx, dialCancel := context.WithTimeout(ctx, c.dialTimeout)
	defer dialCancel()
	conn, err := grpc.DialContext(
		dialCtx,
		c.endpoint,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithBlock(),
	)
	if err != nil {
		return nil, fmt.Errorf("dial auth authority failed: %w", err)
	}
	defer func() {
		_ = conn.Close()
	}()

	challengeReq, err := structpb.NewStruct(map[string]any{
		"entity_type": req.EntityType,
		"entity_id":   req.EntityID,
		"key_id":      strings.TrimSpace(req.KeyID),
		"audience":    strings.TrimSpace(req.Audience),
		"ttl_sec":     float64(60),
	})
	if err != nil {
		return nil, fmt.Errorf("build bootstrap challenge request failed: %w", err)
	}

	challengeResp := &structpb.Struct{}
	challengeCtx, challengeCancel := context.WithTimeout(ctx, c.callTimeout)
	err = conn.Invoke(challengeCtx, bootstrapInitMethodPath, challengeReq, challengeResp)
	challengeCancel()
	if err != nil {
		return nil, fmt.Errorf("bootstrap challenge rpc failed: %w", err)
	}

	challengeEnvelope := challengeResp.AsMap()
	challengeMap := readMap(challengeEnvelope, "challenge")
	if len(challengeMap) == 0 {
		challengeMap = challengeEnvelope
	}
	challengeID := readString(challengeMap, "challenge_id")
	challengeKeyID := readString(challengeMap, "key_id")
	if challengeID == "" {
		return nil, fmt.Errorf("bootstrap challenge response missing challenge_id")
	}
	if challengeKeyID == "" {
		challengeKeyID = strings.TrimSpace(req.KeyID)
	}

	authReq, err := structpb.NewStruct(map[string]any{
		"challenge": challengeMap,
		"signed": map[string]any{
			"challenge_id":        challengeID,
			"key_id":              challengeKeyID,
			"signature_algorithm": "ed25519",
			"signature":           base64.StdEncoding.EncodeToString([]byte("bootstrap:" + challengeID)),
			"signed_at_ms":        float64(time.Now().UnixMilli()),
		},
		"scopes":                   []any{"service:bootstrap"},
		"role":                     "service",
		"require_downstream_token": false,
	})
	if err != nil {
		return nil, fmt.Errorf("build bootstrap authenticate request failed: %w", err)
	}

	authResp := &structpb.Struct{}
	authCtx, authCancel := context.WithTimeout(ctx, c.callTimeout)
	err = conn.Invoke(authCtx, bootstrapAuthMethodPath, authReq, authResp)
	authCancel()
	if err != nil {
		return nil, fmt.Errorf("bootstrap authenticate rpc failed: %w", err)
	}

	authMap := authResp.AsMap()
	stage := normalizeBootstrapStage(readString(authMap, "stage"))
	if stage == "" {
		return nil, fmt.Errorf("bootstrap authenticate response missing stage")
	}

	return &BootstrapHandshakeResult{
		Stage:           stage,
		ActiveCommKeyID: readString(authMap, "active_comm_key_id"),
	}, nil
}

func readString(m map[string]any, key string) string {
	if len(m) == 0 {
		return ""
	}
	v, ok := m[key]
	if !ok || v == nil {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return ""
	}
	return strings.TrimSpace(s)
}

func readMap(m map[string]any, key string) map[string]any {
	if len(m) == 0 {
		return nil
	}
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	child, ok := v.(map[string]any)
	if !ok {
		return nil
	}
	return child
}

func normalizeBootstrapStage(raw string) string {
	stage := strings.TrimSpace(strings.ToLower(raw))
	switch stage {
	case "ready", "bootstrap_stage_ready", "4":
		return "ready"
	case "uninitialized", "bootstrap_stage_uninitialized", "1":
		return "uninitialized"
	case "challenging", "bootstrap_stage_challenging", "2":
		return "challenging"
	case "authenticating", "bootstrap_stage_authenticating", "3":
		return "authenticating"
	default:
		return stage
	}
}
